src/native/containers/dn-queue.h

```diff
@@ -0,0 +1,150 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifndef __DN_QUEUE_H__
+#define __DN_QUEUE_H__
+
+#include "dn-list.h"
+
+typedef void (DN_CALLBACK_CALLTYPE *dn_queue_dispose_func_t) (void *data);
+
+typedef struct _dn_queue_t dn_queue_t;
+struct _dn_queue_t {
+	struct {
+		dn_list_t list;
+	} _internal;
+	uint32_t size;
+};
+
+dn_queue_t *
+_dn_queue_alloc (dn_allocator_t *allocator);
+
+bool
+_dn_queue_init (
+	dn_queue_t *queue,
+	dn_allocator_t *allocator);
+
+static inline dn_queue_t *
+dn_queue_custom_alloc (dn_allocator_t *allocator)
+{
+	return _dn_queue_alloc (allocator);
+}
+
+static inline dn_queue_t *
+dn_queue_alloc (void)
+{
+	return _dn_queue_alloc (DN_DEFAULT_ALLOCATOR);
+}
+
+void
+dn_queue_custom_free (
+	dn_queue_t *queue,
+	dn_queue_dispose_func_t dispose_func);
+
+static inline void
+dn_queue_free (dn_queue_t *queue)
+{
+	dn_queue_custom_free (queue, NULL);
+}
+
+static inline bool
+dn_queue_custom_init (
+	dn_queue_t *queue,
+	dn_allocator_t *allocator)
+{
+	return _dn_queue_init (queue, allocator);
+}
+
+static inline bool
+dn_queue_init (dn_queue_t *queue)
+{
+	return _dn_queue_init (queue, DN_DEFAULT_ALLOCATOR);
+}
+
+void
+dn_queue_custom_dispose (
+	dn_queue_t *queue,
+	dn_queue_dispose_func_t dispose_func);
+
+static inline void
+dn_queue_dispose (dn_queue_t *queue)
+{
+	dn_queue_custom_dispose (queue, NULL);
+}
+
+static inline void **
+dn_queue_front (dn_queue_t *queue)
+{
+	DN_ASSERT (queue && queue->size != 0);
+	return dn_list_front (&queue->_internal.list);
+}
+
+#define dn_queue_front_t(queue, type) \
+	(type *)dn_queue_front ((queue))
+
+static inline void **
+dn_queue_back (dn_queue_t *queue)
+{
+	DN_ASSERT (queue && queue->size != 0);
+	return dn_list_back (&queue->_internal.list);
+}
+
+#define dn_queue_back_t(queue, type) \
+	(type *)dn_queue_back ((queue))
+
+static inline bool
+dn_queue_empty (dn_queue_t *queue)
+{
+	DN_ASSERT (queue);
+	return queue->size == 0;
+}
+
+static inline uint32_t
+dn_queue_size (dn_queue_t *queue)
+{
+	DN_ASSERT (queue);
+	return queue->size;
+}
+
+static inline bool
+dn_queue_push (
+	dn_queue_t *queue,
+	void *data)
+{
+	DN_ASSERT (queue);
+
+	bool result = dn_list_push_back (&queue->_internal.list, data);
+	if (result)
+		queue->size ++;
+
+	return result;
+}
+
+#define dn_queue_push_t(queue, type, data) \
+	dn_queue_push ((queue),((void *)(ptrdiff_t)(data)))
+
+static inline void
+dn_queue_pop (dn_queue_t *queue)
```

Comment:
Maybe we should also add a `dn_queue_custom_pop` for convenience that takes a dispose func argument.  

```c
dn_queue_custom_pop (dn_queue_t *queue, dn_custom_dispose_func_t dispose_func)
{
    void** front_ref = dn_queue_front(queue);
    if (!front_ref) return;
    dispose_func(*front_ref);
    dn_queue_pop(queue);
}
```

Reply:
Thought about that, but then I didn't want to add yet another set of custom functions for that since you could just do front->dispose->pop, so I didn't add it to the pop functions, but we can of course add it for completion meaning that all that internally can dispose an item will have a function with a custom dispose function.